server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java

```diff
@@ -0,0 +1,253 @@
+            }
+            int maxMapSize = maxMapSize(state.limit);
+            ramAccounting.addBytes(calculateRamUsage(maxMapSize));
+        }
+        state.itemsSketch.update(value);
+        return state;
+    }
+
+    private static int maxMapSize(int x) {
+        // max map size should be 4 * the limit based on the power of 2 to avoid errors
+        return (int) Math.pow(2, Math.ceil(Math.log(x) / Math.log(2))) * 4;
+    }
+
+    private static long calculateRamUsage(long maxMapSize) {
+        // The internal memory space usage of item sketch will never exceed 18 * maxMapSize bytes, plus a small
+        // constant number of additional bytes.
+        // https://datasketches.apache.org/docs/Frequency/FrequentItemsOverview.html
+        return maxMapSize * 18L;
+    }
+
+    @Override
+    public TopKState reduce(RamAccounting ramAccounting, TopKState state1, TopKState state2) {
+        return new TopKState(state1.itemsSketch.merge(state2.itemsSketch), state1.dataType, state1.limit);
+    }
+
+    @Override
+    public List<Object> terminatePartial(RamAccounting ramAccounting, TopKState state) {
+        ItemsSketch.Row<Object>[] frequentItems = state.itemsSketch.getFrequentItems(ErrorType.NO_FALSE_NEGATIVES);
+        int limit = Math.min(frequentItems.length, state.limit);
+        var result = new ArrayList<>(limit);
+        for (int i = 0; i < limit; i++) {
+            var item = frequentItems[i];
+            var row = Map.of("item", item.getItem(), "frequency", item.getEstimate());
+            result.add(row);
+        }
+        return result;
+    }
+
+    public DataType<?> partialType() {
+        return TopKStateType.INSTANCE;
+    }
+
+    private static final TopKState EMPTY_STATE = new TopKState(null, DataTypes.NOT_SUPPORTED, -1);
+
+    public record TopKState(@Nullable ItemsSketch<Object> itemsSketch, DataType<?> dataType, int limit) { }
+
+    static final class TopKStateType extends DataType<TopKState> implements Streamer<TopKState> {
+
+        public static final int ID = 4232;
+        public static final TopKStateType INSTANCE = new TopKStateType();
+
+        @Override
+        public int id() {
+            return ID;
+        }
+
+        @Override
+        public Precedence precedence() {
+            return Precedence.CUSTOM;
+        }
+
+        @Override
+        public String getName() {
+            return "topk_state";
+        }
+
+        @Override
+        public Streamer<TopKState> streamer() {
+            return this;
+        }
+
+        @Override
+        public TopKState sanitizeValue(Object value) {
+            return (TopKState) value;
+        }
+
+        @Override
+        @SuppressWarnings({"rawtypes", "unchecked"})
+        public TopKState readValueFrom(StreamInput in) throws IOException {
+            DataType<?> datatype = DataTypes.fromStream(in);
+            int limit = in.readInt();
+            SketchStreamer streamer = new SketchStreamer(datatype.streamer());
```

Comment:
I also tried a typed version with generics, but this just makes everything a lot more complicated without any real benefit since the input value is always object.

Reply:
Everything else is running against `<Object>` just here the raw types makes it really easy to handle the serialization.